### DISCUSSION TITLE
Best or proper way to pass python object into pytest

### DISCUSSION THREAD

**Discussion by AndrewFalanga** [pytest-dev/pytest#9117] [Q&A] [answered]
Hi,

I'm invoking pytest via the `pytest.main(...)` call.  From the invoking python, I need to pass an object reference into the tests so that they may make use of the functions.  In condensed form, this is what I'm doing now:

```
# File: object_reference.py

class ObjRef:
    def __init__(self, obj):
        find_name = lambda o, n: getattr(obj, n, self.throw_up) if obj else self.throw_up
        self.find_data = find_name(obj, 'find_data')
        ...
    def throw_up(self):
        raise NotImplementedError

# then: in main test code
import pytest
import object_reference

class TestExecutor:
    # has lots of methods and properties
    def __init__(self):
        self.comm = SomeCommThing()
    
    def execute(self):
        ref = object_reference.ObjRef(self.comm)
        pytest.main([...], plugins = [ref])
```
As you can see, the `ObjRef` object isn't truly a plugin.  This truly is the simplest way of passing an object with a list of callbacks that the tests make use of (as opposed to somehow making the tests import and instantiate).  However, is there a better method?  I've spent much time Googling and this is the best I've found thus far.  "Callable modules" aren't really a thing without a lot of black magic programming.

**Comment by RonnyPfannschmidt** [comment]
The simplest way to pass a object in is a plugin which provides a fixture

Plugins can be a instance of a class

Methods can be marked as fixtures

**Comment by nicoddemus** [comment, marked as answer]
To exemplify what @RonnyPfannschmidt said, something like this:

```python
# then: in main test code
import pytest
import object_reference

@attr.define 
class MyPlugin:
   _ref: object_reference.ObjRef

   @pytest.fixture
   def ref(self) -> object_reference.ObjRef:
       return self._ref

class TestExecutor:
    # has lots of methods and properties
    def __init__(self):
        self.comm = SomeCommThing()
    
    def execute(self):
        ref = object_reference.ObjRef(self.comm)
        pytest.main([...], plugins = [MyPlugin(ref)])
```

Now your tests have access to a `ref` fixture.

**Reply by tilak-trinity** [reply]
Thanks for this detailed answer:

I have a similar use case- Wherein I need to pass a function during pytest call. below is my implementation. **I get an error _wrapper not defined_**
Any help is appreciated. Below is the reference code


**File main.py**

```
import method_x

#define variables  a , b

class MyPlugin:               
        @pytest.fixture
         def wrapper():
               obj=method_x(var_1= a, var_2= b)
               return obj 

        
print("testing starts")
ref= MyPlugin(wrapper)
```

##  Invoking pytest
` pytest.main(plugins=[ref]) `
